Create Publish Messages to Streams topic (#340)
Macrometacorp/docs: docs/streams/publish-messages.md

---
sidebar_position: 60
title: Publish Messages to Streams
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> This page explains how to publish messages to a stream in Macrometa. Before publishing messages, you must first [create a stream](create-streams.md) to publish the messages to.

Reviewer: ```suggestion
This page explains how to publish messages to a stream in Macrometa. Before publishing messages, you must first [create a stream](create-streams.md) to which you want to publish messages.
```

Probably not a necessary change, but I prefer to avoid ending sentences with "to"

Reviewer 2: The stream will be created automatically if does not exist yet. So, the step `create-streams.md` is redundant.

        if user_input == '0':
            break
        producer.send(user_input)

sendData()
```

</TabItem>
<TabItem value="js" label="JavaScript SDK">

You must [Install the JavaScript SDK](../sdks/install-sdks.md) before you can run this code.

```js
// Connect to GDN.
const jsc8 = require("jsc8");
const client = new jsc8({url: "https://gdn.paas.macrometa.io", apiKey: "XXXXX", fabricName: "_system"});
console.log("Authentication done!!...");

const stream = "streamQuickstart";

async function sendData () {
  console.log("\n ------- Publish Messages  ------");
  const producer = await client.createStreamProducer(stream);

  producer.on("open", () => {
    for (let i = 0; i < 10; i++) {
      const msg1 = `Persistent hello from (${JSON.stringify(i)})`;
      const data = {
        payload: Buffer.from(msg1).toString("base64")
      };

      console.log(`Stream: ${msg1}`);
      producer.send(JSON.stringify(data));
    }
  });
  producer.onclose = function (e) {
    console.log("Closed WebSocket:Producer connection for " + streamName);
  };
}

sendData()
```

</TabItem>
<TabItem value="api-py" label="REST API - Python">

> Use our interactive API Reference with code generation in 18 programming languages to [Publish a Message](https://macrometa.com/docs/api#/operations/PublishStreamMessage).

Reviewer: ```suggestion
Use our interactive API reference with code generation in 18 programming languages to [publish a message](https://macrometa.com/docs/api#/operations/PublishStreamMessage).
```

```py
import requests

# Constants
URL = "api-gdn.paas.macrometa.io"
HTTP_URL = f"https://{URL}"
FABRIC = "_system"
STREAM_NAME = "streamQuickstart"
API_KEY = "XXXXX" # Use your API key here

> AUTH_TOKEN = f"apikey {API_KEY}" # apikey keyword needs to be appended

Reviewer: ```suggestion
AUTH_TOKEN = f"apikey {API_KEY}" # Append the key word for the API key
```

session.headers.update({"content-type": 'application/json'})
session.headers.update({"authorization": AUTH_TOKEN})

# Publish messages
# Send message in body
producerurl = f"wss://{URL}/_ws/ws/v2/producer/persistent/{TENANT_NAME}/{stream_type}.{FABRIC}/{stream_type}s.{STREAM_NAME}"

# Enter your message here
msg = "Hello World"
def create_producer():
    ws = create_connection(producerurl, header=[f"Authorization: {AUTH_TOKEN}"])
    payload = {
        "payload": base64.b64encode(
            six.b(msg)
        ).decode("utf-8")
    }
    ws.send(json.dumps(payload))
    print(f"Message sent: {msg}")
    time.sleep(3)
    response = json.loads(ws.recv())
    if response['result'] == 'ok':
        print("Received acknowledgement that message was delivered successfully")
    else:
        print(f"Failed to publish message: {response}")
    ws.close()

# Or
# Use publish message api to publish message
#url = f"{HTTP_URL}/_fabric/{FABRIC}/_api/streams/{stream_type}s.{STREAM_NAME}/publish?global={IS_GLOBAL}"
#resp = session.post(url, data="Hello")
#print("\nMessage Posted: ", resp.text)
```

</TabItem>
<TabItem value="api-js" label="REST API - JavaScript">

> Use our interactive API Reference with code generation in 18 programming languages to [Publish a Message](https://macrometa.com/docs/api#/operations/PublishStreamMessage).

Reviewer: ```suggestion
Use our interactive API reference with code generation in 18 programming languages to [publish a message](https://macrometa.com/docs/api#/operations/PublishStreamMessage).
```

```js
class APIRequest {
  _headers = {
    Accept: "application/json",
    "Content-Type": "application/json"
  };

  constructor (url, apiKey) {
    this._url = url;

>     this._headers.authorization = `apikey ${apiKey}`; // apikey keyword needs to be appended

Reviewer: ```suggestion
    this._headers.authorization = `apikey ${apiKey}`; // Append the key word for the API key
```
